doc/design/requirements.md

```diff
@@ -0,0 +1,144 @@
+  - if the constraints cannot be satisfied by any installed version, TiUp will show an error
+* The user may specify overrides of a component
+  - any component may be overridden
+  - an override may apply to any version, a specific versions, or a range of versions
+  - an override may use another version or another component (of the same or different version) or a local binary
+  - overrides apply to ways of running a component
+  - when the user runs a component explicitly, the user can cause the override to be ignored
+
+### Third-party components
+
+* Users should be able to publish their own components to public and private servers
+* Users can install third-party components and use them like the standard components
+* TiUp should manage version compatibility between components
+* Users should be able to search for components
+* Users who have published a component should be able to hide it (but not delete it)
+
+### Self-update
+
+* Update TiUp from remote and local servers
+* Uninstall TiUp
+
+### Deployment and managing a deployed cluster
+
+TODO
+
+Better error messages when setting up and managing a cluster
+
+### Sources
+
+Also called mirrors and servers
+
+* PingCAP will host a publicly available server
+* Any third party may host a public or private server
+  - Users must opt-in to using a source
+  - Sources are strictly ordered for a given user (and can be changed by the user)
+  - Sources cannot be specified per-component
+  - All sources are checked in turn for a component
+  - If updating a component causes it to be installed from a different source, the user should be prompted
+* The user may create a local server, that server is created empty
+* Components can be cloned from any source to a local source, either a specific version, all versions, or a range of versions
+* TiUp can manage a server (possibly with a non-standard component); users should not have to manually edit any part of a server
+* Hosts of servers are responsible for any authentication, TiUp will not keep servers private but will provide API hooks for hosts to do so
+
+### Supported products
+
+* TiDB
+* TiKV (with and without TiDB)
+* TiFlash
+* Binlog
+* Blackbox probe
+* Prometheus and Grafana
+
+TiUp should be extensible with user-provided components.
+
+### Interfaces
+
+This section is getting beyond requirements and should be removed, but might be useful for deriving requirements
+
+* SQL client
+* playground
+* cluster
+
+### Non-functional requirements
+
+* Security - we must be able to securely download files and manage clusters since the DB is likely to be security-critical for users. Download binaries is likely to be over the internet. Cluster management might include cross-data center communication on a public network. No need to protect data at rest.
+* Reliability - could bugs in cluster management cause severe issues? That might be important. I think delaying update is not critical as long as it is not for too long, perhaps <1hr?
```

Comment:
I think this may be not much a critical issue, many users trends to delay upgrading of the software in production, especially infrastructures like DB server.